### PR TITLE
Add Armeria as a backend server

### PR DESCRIPTION
**Pull request by ikhoon**
Hi all.
I'm one of the [Armeria](https://armeria.dev/) project maintainers, and also a big fan of http4s project.
Armeria is a Netty based asynchronous microservice framework that supports REST, gRPC, and Thrift.
Armeria is going to support other JVM languages. See https://github.com/line/armeria/issues/1078
I and our users want to use http4s with Armeria. 😀

By supporting Armeria as a backend server, http4s users can do:
- Run gRPC stub such as ScalaPB on the http4s server with sharing a single port.
- Supports various protocols H1, H1C, H2, H2C, HA Proxy, and so on. See https://armeria.dev/docs#features

If this PR is acceptable, I'm willing to add Armeria client integration too in the following PR.

### CODE REVIEW

**Comment by ngbinh** [comment]
this is awesome!

**Comment by rossabaker** [comment]
Thanks, @ikhoon. This is neat.  

I would propose hosting this in either http4s/http4s-armeria or line/http4s-armeria.  The trend has been toward introducing new integration modules in auxiliary repos (e.g., http4s-netty, http4s-finagle), and we have tentative plans to split off some modules already in this repo.  When integrations split off, they can release whenever either http4s _or_ the integrated module make a breaking change, which is nice.  It's also good for the ecosystem if we can reduce core repo dependencies to support new Scala versions faster.  (Java dependencies like armeria aren't so problematic, but this is a definite motivator elsewhere.)

If you'd like to host this in the http4s org, I'll set up a repo.  Or if you think it fits better in another org/account, we'll make sure to link it.

**Comment by rossabaker** [comment]
Also, I'd like to give this a good review, as it's thoughtful work, but let's figure out where it's going to live first.

**Comment by ikhoon** [comment]
> I would propose hosting this in either http4s/http4s-armeria or line/http4s-armeria.

That's is a good idea. Let me talk about which one is better with other Armeria maintainers.

> Also, I'd like to give this a good review

Thank you!

File: armeria-server/src/main/scala/org/http4s/server/armeria/ArmeriaHttp4sHandler.scala [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        .asScala
+        .map(entry => Header(entry.getKey.toString(), entry.getValue))
+        .toList
+    )
+
+  /** Converts a HTTP payload to [[EntityBody]]. */
+  private def toBody(req: HttpRequest): EntityBody[F] =
+    req
+      .toStream[F]
+      .filter(_.isInstanceOf[HttpData])
```
Comment (hamnis) [inline armeria-server/src/main/scala/org/http4s/server/armeria/ArmeriaHttp4sHandler.scala:152]: can be simplified to 
`collect{x: HttpData => Chunk.bytes(x.array())}.flatMap(Stream.chunk)`

File: armeria-server/src/main/scala/org/http4s/server/armeria/ArmeriaHttp4sHandler.scala [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    val response = Response[F](Status.BadRequest).withEntity(parseFailure.sanitized)
+    F.pure(toHttpResponse(response))
+  }
+
+  /** Converts http4s' [[Response]] to Armeria's [[HttpResponse]]. */
+  private def toHttpResponse(response: Response[F]): HttpResponse = {
+    val headers = Stream(toResponseHeaders(response.headers, response.status.some))
+    val body: Stream[F, HttpData] = response.body.chunks.map { chunk =>
+      val byteChunk = chunk.toBytes
+      HttpData.copyOf(chunk.toBytes.values, byteChunk.offset, byteChunk.length)
```
Comment (hamnis) [inline armeria-server/src/main/scala/org/http4s/server/armeria/ArmeriaHttp4sHandler.scala:90]: chunk.toBytes is called twice here

Comment (ikhoon) [reply armeria-server/src/main/scala/org/http4s/server/armeria/ArmeriaHttp4sHandler.scala:90]: 😱

File: armeria-server/src/main/scala/org/http4s/server/armeria/ArmeriaHttp4sHandler.scala (hunk 1 above)
Comment (ikhoon) [reply armeria-server/src/main/scala/org/http4s/server/armeria/ArmeriaHttp4sHandler.scala]: That's what I want. Thanks!

**Comment by ikhoon** [comment]
> If you'd like to host this in the http4s org, I'll set up a repo.

@rossabaker I prefer host `http4s-armeria` in the http4s org. Because other interop projects are already hosted in there. It would help users find this project easily.

**Comment by rossabaker** [comment]
Great. I created a [new repo](https://github.com/http4s/http4s-armeria/) and invited @ikhoon as a maintainer.

* We can continue the review if this code is PR'ed over there.
* It will need a license.  I'm assuming Apache 2.0 since both "parent" projects are Apache 2.0.
* Code of Conduct and Security Policy are inherited through the [.github repo](https://github.com/http4s/.github) I just created.  First time I've used the feature, and I'm not sure that the GitHub UI makes it obvious that we're covered?
* We'll need a versioning scheme.  This has been discussed a bit on http4s-netty and http4s-finagle with no universal answer.
* I probably forgot some things.  Speak up if you're blocked, and feel free to join us on http4s-dev gitter.

**Comment by ikhoon** [comment]
> Great. I created a new repo and invited @ikhoon as a maintainer.

Checked it. Let me close this PR and open a new PR in http4s-armeria with SBT project set up.

I really appreciate your kind feedback. 🙇‍♂️

**Comment by ikhoon** [comment]
I opened a new PR in http4s/http4s-armeria :-)

**Pull request closed without merging**